ambry-router/src/main/java/com/github/ambry/router/DeleteOperation.java

```diff
@@ -349,6 +359,22 @@ private void onErrorResponse(ReplicaId replicaId, RouterException exception, boo
     }
   }
 
+  /**
+   * Check if we should replicate the blob and retry the operation.
+   * @return true if need to do the retry.
+   */
+  private boolean shouldReplicateBlobAndRetry() {
+    // the conditions to trigger ReplicateBlob and retry the original operation:
+    // 1. the feature is enabled.
+    // 2. and at least one replica returned NOT_FOUND status.
+    // 3. and at least one replica returned successful status.
+    // 4. and error code precedence is over AmbryUnavailable. We don't do replication if there BlobExpired or TooManyRequests.
+    RouterErrorCode errorCode = ((RouterException) operationException.get()).getErrorCode();
+    return (routerConfig.routerRepairWithReplicateBlobOnDeleteEnabled && operationTracker.hasNotFound()
+        && operationTracker.getSuccessCount() > 0 && getPrecedenceLevel(errorCode) >= getPrecedenceLevel(
+        RouterErrorCode.AmbryUnavailable));
```

Comment:
Just curious -- why do we need to retry on any error other than `AmbryUnavailable`?

Reply:
On-demand replication is mostly used to deal with the case that we have replicas returns NOT_FOUND or Unavailable. In this case, we will do the on-demand replication. But if there is some replicas returns BlobExpired or BlobAuthorizationFailure some determined error code, we don't want to do the on-demand replication.